torch/csrc/jit/passes/quantization.cpp

```diff
@@ -676,5 +676,246 @@ TORCH_API script::Module InsertObservers(
+  TORCH_INTERNAL_ASSERT(observer_module,
+                        "Expect observer module " +                     \
+                        observer_name.value() + " in quantizeTensor");
+  // remove observer_module
+  observer_modules_to_remove_.push_back(observer_name.value());
+  // remove observer forward call
+  for (const Use& u: v->uses()) {
+    Node* user = u.user;
+    if (user->kind() == prim::CallMethod &&
+        user->s(c10::attr::name) == "forward" &&
+        user->inputs()[0]->node()->kind() == prim::GetAttr &&
+        user->inputs()[0]->node()->s(attr::name) == observer_name) {
+      // Observer forward call node
+      nodes_to_destroy_.push_back(user);
+      // GetAttr node for observer module
+      nodes_to_destroy_.push_back(user->inputs()[0]->node());
+    }
+  }
+
+  // calculate qparams and insert quant/dequant calls
+  auto calculate_qparams = observer_module.value().get_method("calculate_qparams");
+  IValue qparams = calculate_qparams(std::vector<IValue>());
+  Node* dequant;
+  if (v->node()->kind() == prim::GetAttr && v->node()->s(c10::attr::name) == "weight") {
+    dequant = insertQuantDeQuantCall(v, qparams, at::kQInt8);
+  } else {
+    dequant = insertQuantDeQuantCall(v, qparams, at::kQUInt8, insert_after);
+  }
+  v->replaceAllUsesWith(dequant->output());
+  Node* q = traverseToQuantNode(dequant);
+  TORCH_INTERNAL_ASSERT(q);
+  q->replaceInputWith(dequant->output(), v);
+}
+
+void InsertQuantDeQuant(
+    script::Module& module,
+    const std::string& method_name) {
+  script::Method method = module.get_method(method_name);
+  auto graph = method.graph();
+  TORCH_CHECK(graph != nullptr);
+  auto num_activation_inputs = method.num_inputs();
+  std::vector<Value*> values_to_observe;
+  std::vector<Value*> input_values;
+
+  // For traversing all blocks in the graph including subblocks.
+  std::stack<Block*> blocks_to_visit;
+
+  // Mark observer nodes for inputs so we dont add observers
+  // for observers while traversing graph
+  std::unordered_set<Node*> observer_for_input;
+
+  // Add observer for external input nodes excluding parameters
+  // These are treated as activation as they vary across batches
+  // and need to be observed.
+
+  // prim::Param nodes do not belong to the graph. Hence the Insert
+  // point is the beginning of graph node. This also safe guards against
+  // observing a potentially mutated value due to some in-place operation
+  for (size_t idx = 0; idx < num_activation_inputs; ++idx) {
+    auto& v = graph->inputs()[idx];
+    if (v->type()->isSubtypeOf(TensorType::get())) {
+      input_values.push_back(v);
+    }
+  }
+
+  blocks_to_visit.push(graph->block());
+  while (!blocks_to_visit.empty()) {
+    Block* b = blocks_to_visit.top();
+    blocks_to_visit.pop();
+    for (Node* n : b->nodes()) {
+      // Skip nodes that we don't need to observe, e.g. 'prim::Constant' or
+      // observer nodes
+      if (!outputsNeedToBeObserved(n) || observer_for_input.count(n) != 0) {
```

Comment:
Doesn't seem that we're putting anything into `observer_for_input`. Do we want to? If not, please remove it.

Reply:
we don't need that, will cleanup